### PR TITLE
Andrew Shore P05 VertexShader

### PR DESCRIPTION
**Pull request by shore14**
screenshot 19

### CODE REVIEW

File: definitions.h [hunk 1]
```diff
@@ -213,7 +213,207 @@ class BufferImage : public Buffer2D<PIXEL>
...
+       Matrix scaleMatrix(double x, double y, double z)
+       {
+           Matrix m;
+           m.matrix[0][0] = x;
+           m.matrix[1][1] = y;
+           m.matrix[2][2] = z;
+           return ((*this) * m); 
+       }
+
+       Matrix rotateZMatrix(double degrees)
```
Comment (Napler) [inline definitions.h]: We will want to rotate on the other axes at some point as well

File: definitions.h [hunk 2]
```diff
@@ -213,7 +213,207 @@ class BufferImage : public Buffer2D<PIXEL>
...
+       Matrix operator=(Matrix rhs)
+       {
+          for(int i = 0; i < this->row; i++)
+            for(int j = 0; j < rhs.col; j++)
+            {
+                this->matrix[i][j] = rhs.matrix[i][j];
+            }
 
+        }
+    //    Matrix *scaleMatrix(double scale)
```
Comment (Napler) [inline definitions.h]: no need to keep all this code

File: coursefunctions.h [hunk 3]
```diff
@@ -379,42 +385,66 @@ void TestVertexShader(Buffer2D<PIXEL> & target)
...
+         rotate = rotate.rotateZMatrix(-30);
+         colorUniforms.matrix = rotate;
+       // colorUniforms.matrix = colorUniforms.matrix.rotateZMatrix(45);
         DrawPrimitive(TRIANGLE, target, colorTriangle, colorAttributes, &colorUniforms, &myColorFragShader, &myColorVertexShader);
 
         /*************************************************
          * SCALE-TRANSLATE-ROTATE in left-to-right order
          * the previous transformations concatenated.
          ************************************************/
 		// Your scale-translate-rotation code that integrates with 'colorUniforms', used by 'myColorVertexShader' goes here
```
Comment (Napler) [inline coursefunctions.h]: Try multiplying all of the matrices instead of calling the functions.

File: definitions.h [hunk 4]
```diff
@@ -301,7 +536,18 @@ void DefaultVertShader(Vertex & vertOut, Attributes & attrOut, const Vertex & ve
     vertOut = vertIn;
     attrOut = vertAttr;
 }
+void vShader(Vertex & vertOut, Attributes & attrOut,const Vertex & vertIn, const Attributes & vertAttr, const Attributes & uniforms)
+{
+   Matrix unif = uniforms.matrix;
+   Matrix matrixOut = unif * vertIn; 
+//    matrixOut * vertIn;
+   vertOut.x = matrixOut.matrix[0][0];
```
Comment (Napler) [inline definitions.h]: You need matrix multiplication.

File: coursefunctions.h [hunk 5]
```diff
@@ -379,42 +385,66 @@ void TestVertexShader(Buffer2D<PIXEL> & target)
...
         /******************************************************************
 		 * TRANSLATE (move +100 in the X direction, +50 in the Y direction)
          *****************************************************************/
         // Your translating code that integrates with 'colorUniforms', used by 'myColorVertexShader' goes here
-
-		DrawPrimitive(TRIANGLE, target, colorTriangle, colorAttributes, &colorUniforms, &myColorFragShader, &myColorVertexShader);
+        Matrix trans;
+        trans = trans.transMatrix(30, 20, 1);
+        colorUniforms.matrix = colorUniforms.matrix.transMatrix(100, 50, 1);
+        colorUniforms.matrix = trans;
```
Comment (tho14063) [inline coursefunctions.h]: you reset your colorUniforms matrix here back to the 30, 20, 1 translation

File: coursefunctions.h [hunk 6]
```diff
@@ -379,42 +385,66 @@ void TestVertexShader(Buffer2D<PIXEL> & target)
...
+        colorUniforms.matrix = scale;
         DrawPrimitive(TRIANGLE, target, colorTriangle, colorAttributes, &colorUniforms, &myColorFragShader, &myColorVertexShader);
 
         /**********************************************
          * ROTATE 45 degrees in the X-Y plane around Z
          *********************************************/
         // Your rotation code that integrates with 'colorUniforms', used by 'myColorVertexShader' goes here
-
+         Matrix rotate;
+         rotate = rotate.rotateZMatrix(-30);
```
Comment (tho14063) [inline coursefunctions.h]: your rotate should be positive

**Pull request merged**